libbeat/outputs/elasticsearch/event_encoder.go

```diff
@@ -0,0 +1,138 @@
+	indexSelector    outputs.IndexSelector
+}
+
+type encodedEvent struct {
+	// If err is set, the event couldn't be encoded, and other fields should
+	// not be relied on.
+	err error
+
+	// If deadLetter is true, this event produced an ingestion error on a
+	// previous attempt, and is now being retried as a bare event with all
+	// contents included as a raw string in the "message" field.
+	deadLetter bool
+
+	// timestamp is the timestamp from the source beat.Event. It's only used
+	// when reencoding for the dead letter index, so it isn't strictly needed
+	// but it avoids deserializing the encoded event to recover one field if
+	// there's an ingestion error.
+	timestamp time.Time
+
+	id       string
+	opType   events.OpType
+	pipeline string
+	index    string
+	encoding []byte
+}
+
+func newEventEncoderFactory(
+	escapeHTML bool,
+	indexSelector outputs.IndexSelector,
+	pipelineSelector *outil.Selector,
+) queue.EncoderFactory {
+	return func() queue.Encoder {
+		return newEventEncoder(escapeHTML, indexSelector, pipelineSelector)
+	}
+}
+
+func newEventEncoder(escapeHTML bool,
+	indexSelector outputs.IndexSelector,
+	pipelineSelector *outil.Selector,
+) queue.Encoder {
+	buf := bytes.NewBuffer(nil)
+	enc := eslegclient.NewJSONEncoder(buf, escapeHTML)
+	return &eventEncoder{
+		buf:              buf,
+		enc:              enc,
+		pipelineSelector: pipelineSelector,
+		indexSelector:    indexSelector,
+	}
+}
+
+func (pe *eventEncoder) EncodeEntry(entry queue.Entry) (queue.Entry, int) {
+	e, ok := entry.(publisher.Event)
+	if !ok {
+		// Currently all queue entries are publisher.Events but let's be cautious.
+		return entry, 0
+	}
+
+	encodedEvent := pe.encodeRawEvent(&e.Content)
+	e.EncodedEvent = encodedEvent
+	e.Content = beat.Event{}
+	return e, len(encodedEvent.encoding)
+}
+
+// Note: we can't early-encode the bulk metadata that goes with an event,
+// because it depends on the upstream Elasticsearch version and thus requires
+// a live client connection. However, benchmarks show that even for a known
+// version, encoding the bulk metadata and the event together gives slightly
+// worse performance, so there's no reason to try optimizing around this
+// dependency.
+func (pe *eventEncoder) encodeRawEvent(e *beat.Event) *encodedEvent {
+	opType := events.GetOpType(*e)
+	pipeline, err := getPipeline(e, pe.pipelineSelector)
+	if err != nil {
+		return &encodedEvent{err: fmt.Errorf("failed to select event pipeline: %w", err)}
+	}
+	var index string
+	if pe.indexSelector != nil {
+		index, err = pe.indexSelector.Select(e)
+		if err != nil {
+			return &encodedEvent{err: fmt.Errorf("failed to select event index: %w", err)}
+		}
+	}
+
+	id, _ := events.GetMetaStringValue(*e, events.FieldMetaID)
+
+	err = pe.enc.Marshal(e)
+	if err != nil {
+		return &encodedEvent{err: fmt.Errorf("failed to encode event for output: %w", err)}
+	}
+	bufBytes := pe.buf.Bytes()
+	bytes := make([]byte, len(bufBytes))
+	copy(bytes, bufBytes)
+	return &encodedEvent{
```

Comment:
question.  After we copy should we do `pe.buf.Reset()`?  The underlying storage is retained so it shouldn't be a performance hit.  It would signal that we are "done" with the buffer.

also, do we know if it is faster to use `copy` or to do something below?  I haven't tested, just curious.

```
newBuf := bytes.NewBuffer(make([]byte, pe.buf.Len())
io.Copy(newBuf, pe.buf)

and then in encodedEvent

encoding: newBuf.Bytes()
```